lsp/core/aws-lsp-core/src/content/cache/uriCacheRepository.ts

```diff
@@ -0,0 +1,156 @@
+ * hashed file name that contains the cached content.
+ */
+export class UriCacheRepository {
+    private cachedUrisDirPath: string
+    private cacheMetadataPath: string
+
+    constructor(
+        cacheDirRoot: string = DefaultUriCacheLocation.path,
+        private readonly time: TimeProvider = new TimeProvider()
+    ) {
+        this.cachedUrisDirPath = path.join(cacheDirRoot, 'cachedUris')
+        this.cacheMetadataPath = path.join(this.cachedUrisDirPath, 'metadata')
+    }
+
+    /**
+     * Loads metadata for all cached URIs
+     */
+    private async loadCacheMetadata(): Promise<UriCacheMetadata> {
+        await this.ensureCacheDirectoryExists()
+
+        if (!(await exists(this.cacheMetadataPath))) {
+            return {}
+        }
+
+        return JSON.parse(await readFile(this.cacheMetadataPath, { encoding: 'utf-8' }))
+    }
+
+    public async cacheContent(uri: URI, content: string, eTag?: string): Promise<void> {
+        const contentFileName = this.hashUri(uri)
+        const metadataEntry: UriCacheMetadataEntry = {
+            contentFileName: contentFileName,
+            eTag,
+            lastUpdated: this.time.currentMilliseconds(),
+        }
+
+        const allMetadata = await this.loadCacheMetadata()
+        // Update URI cache with response data
+        allMetadata[uri.toString()] = metadataEntry
+        await this.writeContentFileText(contentFileName, content)
+        await this.writeMetadata(allMetadata)
+    }
+
+    public async getContentETag(uri: URI): Promise<string | undefined> {
+        const contentMetadata = await this.getContentMetadata(uri)
+        return contentMetadata?.eTag
+    }
+
+    public async getContentMetadata(uri: URI): Promise<ContentMetadata | undefined> {
+        const cacheMetadata: UriCacheMetadata = await this.loadCacheMetadata()
+        return cacheMetadata[uri.toString()]
+    }
+
+    /** Sets the last updated time to the current time in the `metadata` file.  */
+    public async touchLastUpdatedTime(uri: URI): Promise<void> {
+        const cacheMetadata: UriCacheMetadata = await this.loadCacheMetadata()
+        cacheMetadata[uri.toString()].lastUpdated = this.time.currentMilliseconds()
+        await this.writeMetadata(cacheMetadata)
+    }
+
+    /** Updates the 'metadata' file with the given input */
+    private async writeMetadata(uriCache: UriCacheMetadata): Promise<void> {
+        await this.ensureCacheDirectoryExists()
+        await writeFile(this.cacheMetadataPath, JSON.stringify(uriCache))
+    }
+
+    public async getContent(uri: URI): Promise<string | undefined> {
+        const contentFileName = this.hashUri(uri)
+        return this.getContentFileText(contentFileName)
+    }
+
+    /** Gets the content of a uri from the cache */
+    private async getContentFileText(contentFileName?: string): Promise<string | undefined> {
+        if (contentFileName === undefined) {
+            return
+        }
+
+        await this.ensureCacheDirectoryExists()
```

Comment:
I wasn't sure if exists would throw when any part of the path doesn't exist. I put a check to guard all file accesses.
I plan to leave this since these get accessed pretty frequently over the lifetime of a language server, so a directory (and its contents) will be pretty quickly re-populated.